[SPARK-12425][STREAMING] DStream union optimisation

gpoulin opened the pull request:
Use PartitionerAwareUnionRDD when possbile for optimizing shuffling and
preserving the partitioner.

gpoulin committed:
DStream union optimisation

Use PartitionerAwareUnionRDD when possbile for optimizing shuffling and
preserving the partitioner.

AmplabJenkins commented:
Can one of the admins verify this patch?

zsxwing commented:
Could you create a JIRA and add the JIRA no to the title?

zsxwing commented:
Jenkins, test this please

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/UnionDStream.scala:
```diff
@@ -46,7 +45,11 @@ class UnionDStream[T: ClassTag](parents: Array[DStream[T]])
     }
     if (rdds.size > 0) {
-      Some(new UnionRDD(ssc.sc, rdds))
+      if(rdds.forall(_.partitioner.isDefined) && rdds.flatMap(_.partitioner).toSet.size == 1) {
```
You can use `Some(ssc.sc.union(rdds))` directly here

SparkQA commented:
**[Test build #48025 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48025/consoleFull)** for PR 10382 at commit [`839d9ee`](https://github.com/apache/spark/commit/839d9ee77b708f8c7567683a72d1c07775df66dd).

SparkQA commented:
**[Test build #48025 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48025/consoleFull)** for PR 10382 at commit [`839d9ee`](https://github.com/apache/spark/commit/839d9ee77b708f8c7567683a72d1c07775df66dd).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48025/
Test FAILed.

gpoulin committed:
Simplify RDD union

Deduplicate logic to determine if `PartitionerAwareUnionRDD` should be
used instead of `UnionRDD`.

gpoulin commented:
@zsxwing based on your comment, I took the liberty to deduplicate the logic to determine if `UnionRDD` or a `PartitionerAwareUnionRDD`

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #48214 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48214/consoleFull)** for PR 10382 at commit [`acf6f4e`](https://github.com/apache/spark/commit/acf6f4ec90a8970195ec045a36aced8d4193e209).

SparkQA commented:
**[Test build #48214 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48214/consoleFull)** for PR 10382 at commit [`acf6f4e`](https://github.com/apache/spark/commit/acf6f4ec90a8970195ec045a36aced8d4193e209).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48214/
Test FAILed.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #48220 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48220/consoleFull)** for PR 10382 at commit [`acf6f4e`](https://github.com/apache/spark/commit/acf6f4ec90a8970195ec045a36aced8d4193e209).

zsxwing commented:
ping @davies to take a look

SparkQA commented:
**[Test build #48220 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48220/consoleFull)** for PR 10382 at commit [`acf6f4e`](https://github.com/apache/spark/commit/acf6f4ec90a8970195ec045a36aced8d4193e209).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48220/
Test PASSed.

tdas commented:
Would be nice to add a unit test in the WindowOperationsSuite for this. To make sure that windowed RDDs always have the partitioner attached.

zsxwing commented:
@gpoulin could you resolve the conflicts and add a unit test for window as per @tdas's suggestion? Thanks!

AmplabJenkins commented:
Can one of the admins verify this patch?

davies commented:
LGTM, could you fix the conflict?

srowen commented:
@gpoulin are you able to follow up on this or should I take it over?

gpoulin commented:
Sorry, for the delay, I was quite busy lately. I should have sometime over easter to look at this.

srowen commented:
@gpoulin I'll push this through if you'll rebase, or can just update it separately myself

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #54817 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54817/consoleFull)** for PR 10382 at commit [`3bb5ea3`](https://github.com/apache/spark/commit/3bb5ea3007d89a58d9eb1925f3334c7700ab71e3).

SparkQA commented:
**[Test build #54817 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54817/consoleFull)** for PR 10382 at commit [`3bb5ea3`](https://github.com/apache/spark/commit/3bb5ea3007d89a58d9eb1925f3334c7700ab71e3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54817/
Test FAILed.

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #54861 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54861/consoleFull)** for PR 10382 at commit [`3bb5ea3`](https://github.com/apache/spark/commit/3bb5ea3007d89a58d9eb1925f3334c7700ab71e3).

SparkQA commented:
**[Test build #54861 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54861/consoleFull)** for PR 10382 at commit [`3bb5ea3`](https://github.com/apache/spark/commit/3bb5ea3007d89a58d9eb1925f3334c7700ab71e3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54861/
Test FAILed.

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #54871 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54871/consoleFull)** for PR 10382 at commit [`3bb5ea3`](https://github.com/apache/spark/commit/3bb5ea3007d89a58d9eb1925f3334c7700ab71e3).

SparkQA commented:
**[Test build #54871 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54871/consoleFull)** for PR 10382 at commit [`3bb5ea3`](https://github.com/apache/spark/commit/3bb5ea3007d89a58d9eb1925f3334c7700ab71e3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54871/
Test PASSed.

srowen commented:
Merged to master
